File: example/example-java/src/main/java/io/holunda/camunda/bpm/data/example/rest/ApproveOrderTaskController.java
package io.holunda.camunda.bpm.data.example.rest;

import static io.holunda.camunda.bpm.data.CamundaBpmData.builder;
import static io.holunda.camunda.bpm.data.example.process.OrderApproval.*;

import io.holunda.camunda.bpm.data.example.domain.Order;
import java.math.BigDecimal;
import org.camunda.bpm.engine.TaskService;
import org.camunda.bpm.engine.variable.VariableMap;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/task/approve-order")
public class ApproveOrderTaskController {

  private final TaskService taskService;

  public ApproveOrderTaskController(TaskService taskService) {
    this.taskService = taskService;
  }

  @GetMapping("/{taskId}")
  public ResponseEntity<ApproveTaskDto> loadTask(@PathVariable("taskId") String taskId) {
    final Order order = ORDER.from(taskService, taskId).get();
    final BigDecimal orderTotal = ORDER_TOTAL.from(taskService, taskId).get();
    return ResponseEntity.ok(new ApproveTaskDto(order, orderTotal));
  }

  @PostMapping("/{taskId}")
  public ResponseEntity<Void> completeTask(
      @PathVariable("taskId") String taskId, @RequestBody ApproveTaskCompleteDto userInput) {
    VariableMap vars = builder().set(ORDER_APPROVED, userInput.getApproved()).build();
    taskService.complete(taskId, vars);
    return ResponseEntity.noContent().build();
  }
}

File: example/example-kotlin/src/main/kotlin/rest/ApproveOrderTaskController.kt
package io.holunda.camunda.bpm.data.example.kotlin.rest

import io.holunda.camunda.bpm.data.example.kotlin.domain.Order
import io.holunda.camunda.bpm.data.example.kotlin.process.OrderApproval.Variables.ORDER
import io.holunda.camunda.bpm.data.example.kotlin.process.OrderApproval.Variables.ORDER_APPROVED
import io.holunda.camunda.bpm.data.example.kotlin.process.OrderApproval.Variables.ORDER_TOTAL
import io.holunda.camunda.bpm.data.reader.TaskServiceVariableReader
import io.holunda.camunda.bpm.data.writer.VariableMapWriter
import org.camunda.bpm.engine.TaskService
import org.camunda.bpm.engine.variable.Variables.createVariables
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.math.BigDecimal

@RestController
@RequestMapping("/task/approve-order")
class ApproveOrderTaskController(
  private val taskService: TaskService
) {

  @GetMapping("/{taskId}")
  fun loadTask(@PathVariable("taskId") taskId: String): ResponseEntity<ApproveTaskDto> {
    val order = TaskServiceVariableReader.of(ORDER, taskService, taskId).get()
    val orderTotal = TaskServiceVariableReader.of(ORDER_TOTAL, taskService, taskId).get()
    return ResponseEntity.ok(ApproveTaskDto(order, orderTotal))
  }

  @PostMapping("/{taskId}")
  fun completeTask(@PathVariable("taskId") taskId: String, @RequestBody approveTaskComplete: ApproveTaskCompleteDto): ResponseEntity<Void> {
    val vars = createVariables()
    VariableMapWriter.of(ORDER_APPROVED, vars).set(approveTaskComplete.approved)
    taskService.complete(taskId, vars)
    return ResponseEntity.noContent().build()
  }
}

data class ApproveTaskDto(val order: Order, val orderTotal: BigDecimal)

data class ApproveTaskCompleteDto(val approved: Boolean)
